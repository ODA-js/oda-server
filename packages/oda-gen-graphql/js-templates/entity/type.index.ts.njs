<#@ context 'entity' -#>
<#@ requireAs ('entity/type/enums.graphql.njs', 'enums') #>
<#@ requireAs ('entity/type/entry.graphql.njs', 'type') #>
<#@ requireAs ('entity/connections/types.graphql.njs', 'connections.types') #>
<#@ requireAs ('entity/connections/mutations/types.graphql.njs', 'connections.mutation') #>
<#@ requireAs ('entity/connections/mutations/entry.graphql.njs', 'connections.mutation.entry') #>
<#@ requireAs ('entity/mutations/types.graphql.njs', 'mutation.types') #>
<#@ requireAs ('entity/mutations/entry.graphql.njs', 'mutation.entry') #>
<#@ requireAs ('entity/subscriptions/types.graphql.njs', 'subscription.types') #>
<#@ requireAs ('entity/subscriptions/entry.graphql.njs', 'subscription.entry') #>
<#@ requireAs ('entity/query/entry.graphql.njs', 'query.entry') #>
<#@ requireAs ('entity/viewer/entry.graphql.njs', 'viewer.entry') #>

import { common } from 'oda-gen-graphql';

let { fillDefaults, deepMerge } = common.lib;

import { query } from './query/resolver';
import { viewer } from './viewer/resolver';
import { resolver } from './type/resolver';
import { mutation as connectionMutation } from './connections/mutations/resolver';
import { mutation as entityMutation } from './mutations/resolver';
import { subscriptions as entitySubscription, resolver as subscriptionsUnions } from './subscriptions/resolver';

export class #{entity.name} extends common.types.GQLModule {
  constructor(_args) {
    super(_args);
    this._name = '#{entity.name}';
    this._query = fillDefaults(this._query, query);
    this._viewer = fillDefaults(this._viewer, viewer);
    this._resolver = fillDefaults(this._resolver, resolver, subscriptionsUnions);

    this._typeDef = fillDefaults(this._typeDef, {
      'enums': [`#{partial(entity.partials['enums'], 'enums')}`],
      'type': [`#{partial(entity.partials['type'], 'type')}`],
      'mutationTypes': [`#{partial(entity.partials['mutation.types'], 'mutation.types')}`],
      'subscriptionsTypes': [`#{partial(entity.partials['subscription.types'], 'subscription.types')}`],
      'connectionsTypes': [`#{partial(entity.partials['connections.types'], 'connections.types')}`],
      'connectionsMutation': [`#{partial(entity.partials['connections.mutation'], 'connections.mutation')}`],
    });

    this._mutationEntry = fillDefaults(this._mutationEntry, {
      'mutationEntry': [`#{partial(entity.partials['mutation.entry'], 'mutation.entry')}`],
      'connectionsMutationEntry': [`#{partial(entity.partials['connections.mutation.entry'],'connections.mutation.entry')}`],
    });

    this._subscriptionEntry = fillDefaults(this._subscriptionEntry, {
      'subscriptionEntry': [`#{partial(entity.partials['subscription.entry'], 'subscription.entry')}`],
    });

    this._queryEntry = fillDefaults(this._queryEntry, {
      'queryEntry': [`#{partial(entity.partials['query.entry'], 'query.entry')}`],
    });

    this._mutation = fillDefaults(this._mutation, deepMerge(entityMutation, connectionMutation));
    this._subscription = fillDefaults(this._subscription, entitySubscription);

    this._viewerEntry = fillDefaults(this._viewerEntry, {
      'viewerEntry': [`#{partial(entity.partials['viewer.entry'], 'viewer.entry')}`],
    });
  }
}
