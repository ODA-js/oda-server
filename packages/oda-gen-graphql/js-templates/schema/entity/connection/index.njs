<#@ chunks "$$$main$$$" -#>
<#@ alias 'connection-index'#>
<#@ context 'ctx'#>

#{ partial(ctx.types, 'connection-types') }
#{ partial(ctx.mutations.resolver, 'connection-mutations') }
#{ partial(ctx.mutations.types, 'connection-mutations-input-payload') }

<#-chunkStart(`./connections/index.ts`); -#>
import { Schema } from '../../../common';
#{slot('import-connection-index-slot')}

export default new Schema({
  name: '#{ctx.types.name}.connections',
  items: [
    #{slot('export-connection-index-slot')}
  ],
});
