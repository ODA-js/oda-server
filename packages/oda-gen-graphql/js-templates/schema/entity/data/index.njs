<#@ chunks "$$$main$$$" -#>
<#@ alias 'data-index'#>
<#@ context 'ctx'#>

<#- chunkStart(`../../data/${ctx.name}/types/model.ts`); -#>
#{partial(ctx.model,'data/model')}
<#- chunkEnd(); -#>
<#if(ctx.adapter==='mongoose'){#>
<#- chunkStart(`../../data/${ctx.name}/adapter/connector.ts`); -#>
#{partial(ctx.connector,'data/connector/mongoose')}
<#- chunkStart(`../../data/${ctx.name}/adapter/interface.ts`); -#>
#{partial(ctx.connector,'data/connector/interface')}
<#- chunkStart(`../../data/${ctx.name}/adapter/schema.ts`); -#>
#{partial(ctx.schema,'data/schema/mongoose')}
<#- chunkEnd(); -#>
<# } else {#>
<#- chunkStart(`../../data/${ctx.name}/adapter/connector.ts`); -#>
#{partial(ctx.connector,'data/connector/sequelize')}
<#- chunkStart(`../../data/${ctx.name}/adapter/interface.ts`); -#>
#{partial(ctx.connector,'data/connector/interface')}
<#- chunkStart(`../../data/${ctx.name}/adapter/schema.ts`); -#>
#{partial(ctx.schema,'data/schema/sequelize')}
<#- chunkEnd(); -#>
<#}#>

