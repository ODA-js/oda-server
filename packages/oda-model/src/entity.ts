import {
  EntityBasePersistence,
  EntityBaseInternal,
  EntityBaseMetaInfo,
  EntityBaseInput,
  IEntityBase,
  EntityBase,
  EntityBaseOutput,
} from './entitybase';
import { MetaModelType, Nullable, assignValue } from './types';
import { merge } from 'lodash';

export interface IEntity
  extends IEntityBase<
    EntityMetaInfo,
    EntityPersistence,
    EntityInput,
    EntityOutput
  > {
  implements: Set<string>;
  embedded: boolean | Set<string>;
  abstract: boolean;
}

export interface EntityInternal
  extends EntityBaseInternal<EntityMetaInfo, EntityPersistence> {
  implements: Set<string>;
  embedded: boolean | Set<string>;
  abstract: boolean;
}

export interface EntityPersistence extends EntityBasePersistence {}

export interface EntityInput
  extends EntityBaseInput<EntityMetaInfo, EntityPersistence> {
  implements?: string[];
  embedded?: boolean | string[];
  abstract?: boolean;
}

export interface EntityOutput
  extends EntityBaseOutput<EntityMetaInfo, EntityPersistence> {
  implements?: string[];
  embedded?: boolean | string[];
  abstract?: boolean;
}

export interface EntityMetaInfo extends EntityBaseMetaInfo<EntityPersistence> {}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

export class Entity
  extends EntityBase<
    EntityMetaInfo,
    EntityInput,
    EntityInternal,
    EntityPersistence,
    EntityOutput
  >
  implements IEntity {
  public get modelType(): MetaModelType {
    return 'entity';
  }

  constructor(init: EntityInput) {
    super(merge({}, defaultInput, init));
  }

  get implements(): Set<string> {
    return this.$obj.implements;
  }

  get abstract(): boolean {
    return this.$obj.abstract;
  }

  get embedded(): boolean | Set<string> {
    return this.$obj.embedded;
  }

  public updateWith(input: Nullable<EntityInput>) {
    super.updateWith(input);

    assignValue({
      src: this.$obj,
      input,
      field: 'abstract',
      setDefault: src => (src.abstract = false),
    });
    assignValue<EntityInternal, EntityInput, string[]>({
      src: this.$obj,
      input,
      field: 'implements',
      effect: (src, value) => (src.implements = new Set(value)),
      setDefault: src => (src.implements = new Set()),
    });

    assignValue<EntityInternal, EntityInput, boolean | string[]>({
      src: this.$obj,
      input,
      field: 'embedded',
      effect: (src, value) =>
        (src.embedded = typeof value === 'boolean' ? value : new Set(value)),
      setDefault: src => (src.embedded = false),
    });
  }

  public toObject(): EntityOutput {
    return merge({}, super.toObject(), {
      implements: [...this.implements.values()],
      embedded: this.embedded,
      abstract: this.abstract,
    } as Partial<EntityOutput>);
  }
}
