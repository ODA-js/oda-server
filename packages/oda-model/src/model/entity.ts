import clean from '../lib/json/clean';
import { Field } from './field';
import {
  EntityInput,
  EntityJSON,
  EntityStorage,
  IEntity,
  MetaModelType,
} from './interfaces';
import { ModelPackage } from './modelpackage';
import { EntityBase } from './entitybase';

/**
 * 1. тип объекта который входит на updateWith
 * 2. тип объекта который идет на toObject
 * 3. тип объекта который идет на toJSON
 * 3. тип объекта который идет на выходе clone
 */

export class Entity extends EntityBase implements IEntity {
  public modelType: MetaModelType = 'entity';

  protected $obj!: EntityStorage;

  constructor(obj: EntityInput) {
    super(obj);
  }

  get implements(): Set<string> {
    return this.$obj.implements;
  }

  get abstract(): boolean {
    return !!this.$obj.abstract;
  }

  get embedded(): boolean | string[] {
    return this.$obj.embedded instanceof Set
      ? Array.from(this.$obj.embedded)
      : this.$obj.embedded;
  }

  public ensureImplementation(modelPackage: ModelPackage) {
    const newFields: Map<string, Field> = new Map<string, Field>();
    this.implements.forEach(intrf => {
      if (modelPackage.mixins.has(intrf)) {
        const impl = modelPackage.mixins.get(intrf);
        impl.fields.forEach(f => {
          if (!this.fields.has(f.name)) {
            newFields.set(f.name, f);
          }
        });
      }
    });

    if (newFields.size > 0) {
      const update = this.toJSON();
      update.fields.push(...[...newFields.values()].map(f => f.toJSON()));
      this.updateWith(update);
      this.ensureIds(modelPackage);
    }
  }

  public updateWith(obj: EntityInput) {
    if (obj) {
      super.updateWith(obj);

      const result = { ...this.$obj };
      const impl = new Set(obj.implements);
      result.embedded = Array.isArray(obj.embedded)
        ? new Set(obj.embedded)
        : obj.embedded;
      result.abstract = obj.abstract;
      result.implements = impl;
      this.$obj = result;
    }
  }

  public toObject(modelPackage?: ModelPackage) {
    if (!modelPackage) {
      let res = super.toObject();
      return clean({
        ...res,
        implements: [...this.implements],
        embedded: this.embedded,
        abstract: this.abstract,
      });
    } else {
      let modelRelations = modelPackage.relations.get(this.name);
      if (modelRelations) {
        let res = super.toObject();
        return clean({
          ...res,
          embedded: this.embedded,
          abstract: this.abstract,
          implements: [...this.implements].filter(i =>
            modelPackage.mixins.has(i),
          ),
        });
      }
    }
  }
}
