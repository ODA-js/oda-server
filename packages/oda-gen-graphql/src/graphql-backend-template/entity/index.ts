import * as connections from './connections';
import * as data from './data';
import * as mutations from './mutations';
import * as dataPump from './dataPump';
import * as subscriptions from './subscriptions';
import * as query from './query';
import * as viewer from './viewer';
import * as type from './type';
import * as index from './type.index.ts';
import * as UI from './UI';

export {
  connections,
  data,
  mutations,
  dataPump,
  subscriptions,
  query,
  viewer,
  type,
  index,
  UI,
};
