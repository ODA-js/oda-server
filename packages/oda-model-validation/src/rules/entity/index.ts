import pluralName from './pluralName';

export default [new pluralName()];
