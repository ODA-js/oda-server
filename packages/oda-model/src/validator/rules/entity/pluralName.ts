import { IValidationResult } from '../../../interfaces';
import { IEntityContext } from '../../interfaces';
import { Rule } from '../../rules';

export default class implements Rule<IEntityContext> {
  public name = 'entity-plural-name-the-same';
  public description =
    'plural form of entity`s name of entity must be different from its singular form';
  public validate(context: IEntityContext): IValidationResult[] {
    const result: IValidationResult[] = [];
    if (context.entity.name === context.entity.plural) {
      result.push({
        entity: context.entity.name,
        message: this.description,
        result: 'error',
      });
    }
    return result;
  }
}
