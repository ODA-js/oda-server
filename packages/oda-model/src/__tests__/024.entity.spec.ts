import 'jest';
import { Entity } from '../entity';

describe('Entity', () => {
  it('default', () => {
    const res = new Entity({
      name: 'a',
      embedded: true,
      abstract: true,
      implements: ['IA'],
    });
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
});
