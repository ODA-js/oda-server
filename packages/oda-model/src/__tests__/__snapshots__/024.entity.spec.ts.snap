// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity default: toObject 1`] = `
Object {
  "abstract": true,
  "description": undefined,
  "embedded": true,
  "fields": Array [
    Object {
      "args": Object {},
      "defaultValue": undefined,
      "derived": false,
      "description": undefined,
      "entity": "A",
      "identity": true,
      "indexed": true,
      "inheritedFrom": undefined,
      "list": false,
      "metadata": Object {
        "acl": Object {
          "read": Array [],
          "update": Array [],
        },
        "entity": "A",
        "order": -1,
        "persistence": Object {
          "derived": false,
          "identity": true,
          "indexed": true,
          "persistent": true,
          "required": true,
        },
      },
      "name": "id",
      "order": -1,
      "persistent": true,
      "required": true,
      "title": undefined,
      "type": "ID",
    },
  ],
  "implements": Array [
    "IA",
  ],
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "readMany": Array [],
      "readOne": Array [],
      "update": Array [],
    },
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "persistence": Object {
      "indexes": Object {},
    },
    "titlePlural": "As",
  },
  "name": "A",
  "operations": Array [],
  "title": undefined,
}
`;
